core/services/eth/client.go

```diff
@@ -256,7 +250,66 @@ func (client *client) SubscribeFilterLogs(ctx context.Context, q ethereum.Filter
 }
 
 func (client *client) SubscribeNewHead(ctx context.Context, ch chan<- *Head) (ethereum.Subscription, error) {
-	return client.pool.EthSubscribe(ctx, ch, "newHeads")
+	csf := newChainIDSubForwarder(client.ChainID(), ch)
+	err := csf.start(client.pool.EthSubscribe(ctx, csf.srcCh, "newHeads"))
+	if err != nil {
+		return nil, err
+	}
+	return csf, nil
+}
+
+var _ ethereum.Subscription = &chainIDSubForwarder{}
+
+// chainIDSubForwarder wraps a head subscription in order to intercept and augment each head with chainID before forwarding.
+type chainIDSubForwarder struct {
+	chainID *big.Int
+	destCh  chan<- *Head
+	srcCh   chan *Head
+	srcSub  ethereum.Subscription
+	err     chan error
+}
+
+func newChainIDSubForwarder(chainID *big.Int, ch chan<- *Head) *chainIDSubForwarder {
+	return &chainIDSubForwarder{
+		chainID: chainID,
+		srcCh:   make(chan *Head),
+		destCh:  ch,
+		err:     make(chan error),
+	}
+}
+
+// start spawns the forwarding loop for sub.
+func (c *chainIDSubForwarder) start(sub ethereum.Subscription, err error) error {
+	if err != nil {
+		close(c.srcCh)
+		return err
+	}
+	c.srcSub = sub
+	go c.forwardLoop()
+	return nil
+}
+
+// forwardLoop receives from src, adds the chainID, and then sends to dest.
+func (c *chainIDSubForwarder) forwardLoop() {
+	defer close(c.srcCh)
+	for {
+		select {
+		case err := <-c.srcSub.Err():
+			c.err <- err
+			return
+		case h := <-c.srcCh:
+			h.EVMChainID = utils.NewBig(c.chainID)
+			c.destCh <- h
+		}
+	}
+}
```

Comment:
This is the messy part, but I think it's necessary. Open to other ideas.

Reply:
Added another select case in order to synchronize unsubscribing from `forwardLoop()` to prevent sending afterwards.